src/StackExchange.Redis/Configuration/Tunnel.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Pipelines.Sockets.Unofficial;
+
+namespace StackExchange.Redis.Configuration
+{
+    /// <summary>
+    /// Allows interception of the transport used to communicate with Redis.
+    /// </summary>
+    public abstract class Tunnel
+    {
+        /// <summary>
+        /// Gets the underlying socket endpoint to use when connecting to a logical endpoint.
+        /// </summary>
+        /// <remarks><c>null</c> should be returned if a socket is not required for this endpoint.</remarks>
+        public virtual ValueTask<EndPoint?> GetSocketConnectEndpointAsync(EndPoint endpoint, CancellationToken cancellationToken) => new(endpoint);
+
+        /// <summary>
+        /// Allows modification of a <see cref="Socket"/> between creation and connection.
+        /// Passed in is the endpoint we're connecting to, which type of connection it is, and the socket itself.
+        /// For example, a specific local IP endpoint could be bound, linger time altered, etc.
+        /// </summary>
+        public virtual ValueTask BeforeSocketConnectAsync(EndPoint endPoint, ConnectionType connectionType, Socket? socket, CancellationToken cancellationToken) => default;
+
+        /// <summary>
+        /// Invoked on a connected endpoint before server authentication and other handshakes occur, allowing pre-redis handshakes. By returning a custom <see cref="Stream"/>,
+        /// the entire data flow can be intercepted, providing entire custom transports.
+        /// </summary>
+        public virtual ValueTask<Stream?> BeforeAuthenticateAsync(EndPoint endpoint, ConnectionType connectionType, Socket? socket, CancellationToken cancellationToken) => default;
+        /// <inheritdoc/>
+        public abstract override string ToString();
+
+        private sealed class HttpProxyTunnel : Tunnel
+        {
+            public EndPoint Proxy { get; }
+            public HttpProxyTunnel(EndPoint proxy) => Proxy = proxy ?? throw new ArgumentNullException(nameof(proxy));
+
+            public override ValueTask<EndPoint?> GetSocketConnectEndpointAsync(EndPoint endpoint, CancellationToken cancellationToken) => new(Proxy);
+
+            public override async ValueTask<Stream?> BeforeAuthenticateAsync(EndPoint endpoint, ConnectionType connectionType, Socket? socket, CancellationToken cancellationToken)
+            {
+                if (socket is not null)
+                {
+                    var encoding = Encoding.ASCII;
+                    var ep = Format.ToString(endpoint);
+                    const string Prefix = "CONNECT ", Suffix = " HTTP/1.1\r\n\r\n", ExpectedResponse = "HTTP/1.1 200 OK\r\n\r\n";
```

Comment:
@mgravell According to https://datatracker.ietf.org/doc/html/draft-luotonen-web-proxy-tunneling-01#section-3.2, the expected response should be `HTTP/1.1 200 Connection established` instead of `HTTP/1.1 200 OK`. Currently it's not possible to use a HTTP proxy that correctly implements this behavior (e.g. Squid).

Reply:
Created a PR: #2448